stdlib/public/core/String.swift

```diff
@@ -405,15 +405,31 @@ extension String {
   // This force type-casts element to UInt8, since we cannot currently
   // communicate to the type checker that we proved this with our dynamic
   // check in String(decoding:as:).
+  // **Important:** The 'repairsMade' part of the result is not reliable.
   @_alwaysEmitIntoClient
   @inline(never) // slow-path
   private static func _fromNonContiguousUnsafeBitcastUTF8Repairing<
     C: Collection
   >(_ input: C) -> (result: String, repairsMade: Bool) {
     _internalInvariant(C.Element.self == UInt8.self)
-    return Array(input).withUnsafeBufferPointer {
-      let raw = UnsafeRawBufferPointer($0)
-      return String._fromUTF8Repairing(raw.bindMemory(to: UInt8.self))
+    if #available(macOS 10.16, iOS 14.0, watchOS 7.0, tvOS 14.0, *) {
+      let count = input.count
+      let string = String(unsafeUninitializedCapacity: count) { buffer in
+        var iterAndCount = unsafeBitCast(
```

Comment:
I think going to/from a UnsafeRawPointer as in the original code is better than the bit cast. @atrick, thoughts?

Reply:
I agree it's better than the bit cast.

Reply:
Anything is better than `bitcast`. That's about the worst thing you can do with a pointer in terms of danger level and confusing the compiler.

Looks like `UnsafeRawBufferPointer.initializeMemory(as:from:)` would work here as an easy hack (a hack because it erases the String's CodeUnit rather than converting the collection's memory to the statically known type).

Converting Collection's element type from a dynamically known to statically known type is exactly the intended use case for `assumingMemoryBound(to:)`. It isn't available on buffer types, so you'd need to explicitly instantiate UnsafeBufferPointer<UInt8> from the result of calling `assumingMemoryBound` on the collection's raw memory. No `count` adjustment needed here. @glessard can we consider `UnsafeRawBufferPointer.assumingMemoryBound(to:)` as a proposed API cleanup?